test/javascripts/turbolinks_test.coffee

```diff
@@ -1,210 +1,946 @@
+    document.querySelectorAll("link[href=\"#{ASSET_FIXTURES[filename]}\"]")
+
+  queryForScript = (filename) ->
+    document.querySelectorAll("script[src=\"#{ASSET_FIXTURES[filename]}\"]")
+
+  linkOrderPosition = (filename) ->
+    linksInHead = [].slice.call(document.head.children).filter (node) ->
+      node.nodeName == 'LINK'
+    linksInHead.indexOf(queryForLink(filename)[0])
+
+  resetPage = ->
+    document.head.innerHTML = ""
+    document.body.innerHTML = """
+      <div id="turbo-area" refresh="turbo-area"></div>
+    """
+
+  urlFor = (slug) ->
     window.location.origin + slug
 
+  visit = ({options, url}, callback) ->
+    $(document).one('page:load', (event) ->
+      setTimeout((-> callback(event)), 0)
+    )
+    Turbolinks.visit('/' + url, options)
+
   beforeEach ->
-    @server = sinon.fakeServer.create()
+    pushStateStub = stub(Turbolinks, 'pushState')
+    replaceStateStub = stub(Turbolinks, 'replaceState')
+    stub(Turbolinks, 'fullPageNavigate', -> $(document).trigger('page:load'))
+
+    server = sinon.fakeServer.create()
+    Object.keys(ROUTES).forEach (url) ->
+      server.respondWith('/' + url, ROUTES[url])
+    server.autoRespond = true
+
     $("script").attr("data-turbolinks-eval", false)
     $("#mocha").attr("refresh-never", true)
-    @replaceStateStub = stub(Turbolinks, "replaceState")
-    @pushStateStub = stub(Turbolinks, "pushState")
-    document.body.appendChild(createTurboNodule())
+
+    resetPage()
 
   afterEach ->
-    @server.restore()
-    @pushStateStub.restore()
-    @replaceStateStub.restore()
-    $("#turbo-area").remove()
+    Turbolinks.fullPageNavigate.restore()
+    Turbolinks.pushState.restore()
+    Turbolinks.replaceState.restore()
 
-  html_one = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi there!</title>
-      </head>
-      <body>
-        <div>YOLO</div>
-        <div id="turbo-area" refresh="turbo-area">Hi bob</div>
-      </body>
-    </html>
-  """
-
-  script_response = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi</title>
-      </head>
-      <body>
-        <script id="turbo-area" refresh="turbo-area">globalStub()</script>
-      </body>
-    </html>
-  """
-
-  script_response_turbolinks_eval_false = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi</title>
-      </head>
-      <body>
-        <script data-turbolinks-eval="false" id="turbo-area" refresh="turbo-area">globalStub()</script>
-      </body>
-    </html>
-  """
-
-  response_with_refresh_always = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi</title>
-      </head>
-      <body>
-        <div id="div1" refresh="div1">
-          <div id="div2" refresh-always>Refresh-always</div>
-        </div>
-      </body>
-    </html>
-  """
+    $(document).off(TURBO_EVENTS.join(' '))
+    $("#turbo-area").remove()
+    server.restore()
 
   it 'is defined', ->
-    assert Turbolinks
+    assert(Turbolinks)
```

Comment:
The `beforeEach`'s stubbing will fail if `Turbolinks` isn't defined.  Don't think this test is reachable.